### PR TITLE
[flang][OpenMP] Add parsing support for Task detach

### PR DESCRIPTION
**Pull request by NimishMishra**
Add parsing support for task detach, along with parse/unparse tests.

### CODE REVIEW

File: flang/include/flang/Parser/parse-tree.h [hunk 1]
```diff
@@ -3515,6 +3515,12 @@ struct OmpIfClause {
   std::tuple<std::optional<DirectiveNameModifier>, ScalarLogicalExpr> t;
 };
 
+// OpenMPv5.2 12.5.2 detach-clause -> DETACH (event-handle)
+struct OmpDetachClause {
+  WRAPPER_CLASS_BOILERPLATE(OmpDetachClause, OmpObject);
+  CharBlock source;
```
Comment (kiranchandramohan) [inline flang/include/flang/Parser/parse-tree.h:3521]: Is this used currently? Don't you have to use a `sourced()` parser to fill it?

**Review by kiranchandramohan — COMMENTED** [review]
Can you add a TODO in lowering (OpenMP.cpp) to catch that there is no lowering support?

File: flang/lib/Parser/openmp-parsers.cpp [hunk 2]
```diff
@@ -298,6 +298,9 @@ TYPE_CONTEXT_PARSER("Omp LINEAR clause"_en_US,
         construct<OmpLinearClause>(construct<OmpLinearClause::WithoutModifier>(
             nonemptyList(name), maybe(":" >> scalarIntConstantExpr)))))
 
+// 12.5.2 detach-clause -> DETACH (event-handle)
```
Comment (mjklemm) [inline flang/lib/Parser/openmp-parsers.cpp:301]:
```suggestion
// OpenMPv5.2 12.5.2 detach-clause -> DETACH (event-handle)
```

**Review by NimishMishra — COMMENTED** [review]
Thanks all for the reviews.

**Review by tblah — APPROVED** [review]
LGTM, thanks!

**Comment by NimishMishra** [comment]
Can anyone suggest on how to fix the buildbot failure on `openmp-offload-sles-build-only`

Issue: "Cannot read module file for module 'omp_lib': Source file 'omp_lib.mod' was not found"

We need the module file to access `omp_event_handle_kind` while defining the event handle. Is there a way to disable the test on a target when the module file is missing?

**Comment by DavidSpickett** [comment]
FYI there are actually 2 failures on Linaro's bots: https://lab.llvm.org/buildbot/#/builders/29/builds/6009

You won't be getting failure emails because your commit address is set to a GitHub noreply address. Consider changing that to a real email address for future contributions.

**Comment by DavidSpickett** [comment]
@luporl might be able to help here.

On a high level, you would find out what produces that omp_lib (`openmp` being an enabled subproject?) and then add a lit test feature for that thing. Maybe this existing one would do the job?
```
! REQUIRES: openmp_runtime
```

**Comment by kiranchandramohan** [comment]
> Can anyone suggest on how to fix the buildbot failure on `openmp-offload-sles-build-only`
> 
> Issue: "Cannot read module file for module 'omp_lib': Source file 'omp_lib.mod' was not found"
> 
> We need the module file to access `omp_event_handle_kind` while defining the event handle. Is there a way to disable the test on a target when the module file is missing?

Can you try adding `%openmp_flags` to see whether it fixes the issue like in the following?
```
! RUN: %flang_fc1 -emit-hlfir %openmp_flags %s -o - 2>&1 | FileCheck %s
! RUN: bbc %openmp_flags -emit-hlfir -o - %s 2>&1 | FileCheck %s
```

**Comment by NimishMishra** [comment]
Thanks @DavidSpickett and @kiranchandramohan for the suggestions. I think this should solve the issue. I am monitoring the buildbots at the moment, will update here once they are clean. Thanks

**Comment by NimishMishra** [comment]
The buildbots [flang-aarch64-rel-assert](https://lab.llvm.org/buildbot/#/builders/29) and [openmp-offload-sles-build-only](https://lab.llvm.org/buildbot/#/builders/140) are now clean. `%openmp_flags` and `! REQUIRES: openmp_runtime` were sufficient. Thanks all.

**Pull request merged**